src/motion/integrator_utils.F

```diff
@@ -1302,6 +1305,223 @@ SUBROUTINE vv_first(tmp, atomic_kind_set, local_particles, particle_set, &
       END DO
    END SUBROUTINE vv_first
 
+! **************************************************************************************************
+!> \brief First half of the velocity-verlet algorithm : update velocity by half
+!>        step and positions by full step, modified for frozen atoms
+!> \param tmp ...
+!> \param atomic_kind_set ...
+!> \param local_particles ...
+!> \param particle_set ...
+!> \param core_particle_set ...
+!> \param shell_particle_set ...
+!> \param nparticle_kind ...
+!> \param shell_adiabatic ...
+!> \param dt ...
+!> \param lfixd_list ...
+!> \param u ...
+!> \par History
+!>      none
+!> \author Paul Clabaut (Jully 2021)
+! **************************************************************************************************
+   SUBROUTINE vv_first_frozen(tmp, atomic_kind_set, local_particles, particle_set, &
```

Comment:
Maybe instead of duplicating the code you could just add flag to the `vv_first` routine above?